keps/sig-scheduling/5075-dra-consumable-capacity/README.md

```diff
@@ -0,0 +1,1150 @@
+}
+
+
+// CapacitySharingPolicyRange defines a valid range for consumable capacity values.
+//
+// - If the requested amount is less than Minimum, it is rounded up to the Minimum value.
+// - If the requested amount is between Minimum and Maximum, ChunkSize is set,
+//   and the amount does not align with the ChunkSize,
+//   it will be rounded up to the next value matching Minimum + (n * ChunkSize).
+// - If the requested or rounded amount exceeds Maximum (if set), the request does not satisfy the policy,
+//   and the device cannot be allocated.
+//
+// - If ChunkSize is not set, the requested amount is used as-is, provided it falls within the range.
+type CapacitySharingPolicyRange struct {
+    // Minimum specifies the minimum capacity allowed for a consumption request.
+    // This also acts as the default if no value is specified.
+    //
+    // +required
+    Minimum resource.Quantity
+
+    // Maximum defines the upper limit for capacity that can be requested.
+    //
+    // +optional
+    Maximum *resource.Quantity
+
+    // ChunkSize defines the step size between valid capacity amounts within the range.
+    // If set, requested amounts are rounded up to the nearest multiple of ChunkSize from the Minimum.
+    //
+    // +optional
+    ChunkSize *resource.Quantity
+}
+```
+
+#### ResourceClaimSpec's DeviceRequest
+
+```go
+
+type DeviceRequest struct {
+...
+   // CapacityRequests define resource requirements against each capacity.
+   //
+   // +optional
+   // +featureGate=DRAConsumableCapacity
+   CapacityRequests *CapacityRequirements
+}
+
+
+// CapacityRequirements defines the capacity requirements for a specific device request.
+type CapacityRequirements struct {
+    // Minimum defines the minimum amount of each device capacity required for the request.
+    //
+    // If the capacity has a sharing policy, this value is rounded up to the nearest valid amount
+    // according to that policy. The rounded value is used during scheduling to determine how much capacity to consume.
+    //
+    // If the quantity does not have a sharing policy, this value is used as an additional filtering
+    // condition against the available capacity on the device.
+    // This is semantically equivalent to a CEL selector with
+    // `device.quantity[<domain>].<name>.compareTo(<minimum quantity>) >= 0`.
+    //
+    // +optional
+   Minimum map[QualifiedName]resource.Quantity
+}
+
+// DeviceConstraint must have exactly one field set besides Requests.
+type DeviceConstraint struct {
+	...
+   // DistinctAttribute requires that all devices in question have this
+   // attribute and that its type and value are unique across those devices.
+   //
+   // This acts as the inverse of MatchAttribute.
+   //
+   // This constraint is used to avoid allocating multiple requests to the same shareable device
+   // by ensuring attribute-level differentiation.
+   //
+   // +optional
+   // +oneOf=ConstraintType
+   DistinctAttribute *FullyQualifiedName
```

Comment:
Is a single value here sufficient? In the context of this KEP, this seems focused on avoiding matching multiple requests to the same device (so would use some unique device identity attribute, presumably), but are there other scenarios where spreading would be desired that would involve more than one attribute? Normally, topology-type things are multi-dimensional.

It might be worth asking someone from scheduling to take a peek at this bit to see if it is generally useful ... this reminds me a lot of the `{required,preferred} {affinity,anti-affinity}` scheduling directives in pod spec

Reply:
matchAtttribute is a slice, I would expect the same here.

Reply:
Oh, take that back. You list multiple constraints, in order, each with one matchAttribute to do that.

So, you could do the same here. You could even mix them "make sure they are in the same pcieRoot, but not the same physical device" could be expressed with two constraints.

Reply:
ah, ok